File: apps/server/app/usecases/play_track.py
"""Play one track: resolve → probe → spawn → stream chunks → classify.

Extracted from StreamEngine._play_item's attempt body. The runner owns a
single playback attempt and reports facts; the retry policy, interrupt
dispatch, and state mutation stay with the engine (which implements
AttemptHooks). All time access flows through the injected clock, so tests
drive retries and slow-read paths in microseconds.

Never imports adapters: collaborators arrive as ports/duck-typed objects and
engine capabilities behind the AttemptHooks protocol.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass, field
from typing import Protocol, runtime_checkable

from app.domain.outcomes import (
    AttemptFacts,
    AttemptOutcome,
    classify_attempt,
    expected_duration_seconds,
    slow_chunk_read,
)
from app.domain.ports import Clock, ResolvedTrackLike, Transcoder

logger = logging.getLogger(__name__)


class QueueItemLike(Protocol):
    id: int
    source_url: str
    normalized_url: str | None
    duration_seconds: int | None
    title: str | None


@dataclass(frozen=True)
class TrackAttemptRequest:
    queue_item: QueueItemLike
    attempt: int
    default_seek_seconds: float


@dataclass
class TrackAttemptResult:
    outcome: AttemptOutcome
    reason: str | None = None
    chunks_sent: int = 0
    bytes_sent: int = 0
    elapsed_seconds: float = 0.0
    expected_seconds: float = 0.0
    resolved: ResolvedTrackLike | None = None
    seek_seconds: float = 0.0
    stderr_text: str = field(default="")


@runtime_checkable
class AttemptHooks(Protocol):
    """Engine capabilities the attempt needs. Implemented by StreamEngine."""

    def resolve_track(self, item: QueueItemLike, force_refresh: bool) -> ResolvedTrackLike: ...

    def on_resolved_metadata(self, resolved: ResolvedTrackLike) -> None:
        """State updates (thumbnail/channel/is_live) once metadata is known."""

    def mark_item_resolved(self, item_id: int, normalized_url: str) -> None: ...

    def remember_resolved(self, resolved: ResolvedTrackLike) -> None: ...

    def consume_seek(self, default_seconds: float) -> float: ...

    def set_playback_offset(self, seconds: float) -> None: ...

    def get_prefetched_audio(self, item_id: int) -> str | None: ...

    def prefetch_audio(self, item_id: int, source_url: str) -> None: ...

    def uses_direct_ffmpeg(self, item: QueueItemLike) -> bool: ...

    def register_active_process(self, process: object) -> None: ...

    def trigger_prefetch_upcoming(self) -> None: ...

    def notify_state_changed(self) -> None: ...

    def start_transition_silence(self) -> tuple[object, object]: ...

    def stop_transition_silence(self, stop_event: object, worker: object) -> None: ...

    def on_first_chunk(self) -> None: ...

    def write_chunk(self, chunk: bytes) -> None: ...

    def check_interrupt(self) -> str | None:
        """None = keep going; otherwise the interrupt reason ('stop' ends the engine)."""

    def consume_interrupt_reason(self) -> str: ...


class TrackAttemptRunner:
    def __init__(
        self,
        *,
        transcoder: Transcoder,
        clock: Clock,
        chunk_size: int,
    ) -> None:
        self._transcoder = transcoder
        self._clock = clock
        self._chunk_size = chunk_size

    def run(self, request: TrackAttemptRequest, hooks: AttemptHooks) -> TrackAttemptResult:
        item = request.queue_item
        silence_pair: tuple[object, object] | None = None
        started_at = self._clock()
        phase = "resolve"
        try:
            silence_pair = hooks.start_transition_silence()
            resolved = self._resolve(request, hooks)
            phase = "spawn"
            probed_duration = self._probe_duration(resolved)
            hooks.mark_item_resolved(item.id, resolved.normalized_url)
            hooks.remember_resolved(resolved)
            hooks.on_resolved_metadata(resolved)
            seek_seconds = hooks.consume_seek(request.default_seek_seconds)
            hooks.set_playback_offset(seek_seconds)

            process, source_process = self._spawn(request, resolved, seek_seconds, hooks)
            hooks.trigger_prefetch_upcoming()
            hooks.notify_state_changed()
            logger.info("Notifying state changed before first streamed chunk")

            def _on_first_chunk() -> None:
                nonlocal silence_pair
                if silence_pair is not None:
                    hooks.stop_transition_silence(*silence_pair)
                    silence_pair = None
                hooks.on_first_chunk()

            chunks, byte_count, stderr_text = self._stream_chunks(
                item.id, request.attempt, process, source_process, hooks, _on_first_chunk
            )
            elapsed = max(0.0, self._clock() - started_at)

            reason = self._interrupt_or_none(hooks, allow_stop_only=True)
            if reason is not None:
                raise InterruptedError(reason if reason != "stop" else "stop")

            expected = expected_duration_seconds(probed_duration, resolved.duration_seconds, item.duration_seconds)
            verdict = classify_attempt(
                AttemptFacts(
                    ffmpeg_return_code=self._return_code(process),
                    source_return_code=self._return_code(source_process) if source_process is not None else 0,
                    elapsed_seconds=elapsed,
                    expected_seconds=expected,
                    stderr_text=stderr_text,
                )
            )
            return TrackAttemptResult(
                outcome=verdict.outcome,
                reason=verdict.reason,
                chunks_sent=chunks,
                bytes_sent=byte_count,
                elapsed_seconds=elapsed,
                expected_seconds=expected,
                resolved=resolved,
                seek_seconds=seek_seconds,
                stderr_text=stderr_text,
            )
        except InterruptedError:
            raise
        except Exception as exc:  # resolve/probe/spawn failures become retryable outcomes
            outcome = AttemptOutcome.retry_source if phase == "resolve" else AttemptOutcome.retry_ffmpeg
            return TrackAttemptResult(
                outcome=outcome,
                reason=str(exc),
                elapsed_seconds=max(0.0, self._clock() - started_at),
            )
        finally:
            if silence_pair is not None:
                hooks.stop_transition_silence(*silence_pair)

    # ---------------------------------------------------------------- phases

    def _resolve(self, request: TrackAttemptRequest, hooks: AttemptHooks) -> ResolvedTrackLike:
        return hooks.resolve_track(request.queue_item, force_refresh=request.attempt > 1)

    def _probe_duration(self, resolved: ResolvedTrackLike) -> float | None:
        probe_source = getattr(self._transcoder, "probe_source", None)
        try:
            probe = probe_source(resolved.stream_url) if callable(probe_source) else None
            if probe is None:
                return None
            raw = probe.get("duration_seconds")
            return float(raw) if raw is not None else None
        except Exception:
            return None

    def _spawn(
        self,
        request: TrackAttemptRequest,
        resolved: ResolvedTrackLike,
        seek_seconds: float,
        hooks: AttemptHooks,
    ) -> tuple[object, object | None]:
        item = request.queue_item
        spawn_for_source = getattr(self._transcoder, "spawn_for_source", None)
        if not callable(spawn_for_source):
            raise RuntimeError("ffmpeg source playback is unavailable")
        prefetched = hooks.get_prefetched_audio(item.id)
        if not prefetched and not resolved.is_live and not hooks.uses_direct_ffmpeg(item):
            hooks.prefetch_audio(item.id, item.source_url)
            interrupt = hooks.check_interrupt()
            if interrupt is not None:
                raise InterruptedError(hooks.consume_interrupt_reason())
            prefetched = hooks.get_prefetched_audio(item.id)
        if prefetched:
            process = spawn_for_source(prefetched, start_at_seconds=seek_seconds)
        elif seek_seconds > 0 or resolved.is_live or hooks.uses_direct_ffmpeg(item):
            process = spawn_for_source(resolved.stream_url, start_at_seconds=seek_seconds)
        else:
            raise RuntimeError("ffmpeg source playback is unavailable")
        hooks.register_active_process(process)
        return process, None

    def _stream_chunks(
        self,
        item_id: int,
        attempt: int,
        process: object,
        source_process: object | None,
        hooks: AttemptHooks,
        on_first_chunk,
    ) -> tuple[int, int, str]:
        chunks = 0
        byte_count = 0
        while True:
            interrupt = hooks.check_interrupt()
            if interrupt is not None:
                raise InterruptedError(hooks.consume_interrupt_reason() if interrupt != "stop" else "stop")
            read_started = self._clock()
            chunk = self._read_chunk(process)
            read_seconds = self._clock() - read_started
            if chunk and slow_chunk_read(read_seconds):
                logger.warning(
                    "Slow ffmpeg read while streaming track_id=%s attempt=%s chunk_index=%s "
                    "read_seconds=%.3f requested_bytes=%s received_bytes=%s",
                    item_id, attempt, chunks, read_seconds, self._chunk_size, len(chunk),
                )
            if not chunk:
                return chunks, byte_count, self._stderr_text(process, source_process)
            if chunks == 0:
                on_first_chunk()
            hooks.write_chunk(chunk)
            chunks += 1
            byte_count += len(chunk)

    # --------------------------------------------------------------- helpers

    def _read_chunk(self, process: object) -> bytes:
        read_chunk = getattr(self._transcoder, "read_chunk", None)
        stdout = getattr(process, "stdout", None)
        if callable(read_chunk):
            return read_chunk(stdout, self._chunk_size)
        if stdout is None:
            return b""
        return stdout.read(self._chunk_size)

    @staticmethod
    def _stderr_text(process: object, source_process: object | None) -> str:
        def _read(proc: object) -> str:
            pipe = getattr(proc, "stderr", None)
            if pipe is None:
                return ""
            return pipe.read().decode("utf-8", errors="replace").strip()

        return f"{_read(process)}\n{_read(source_process)}".strip()

    @staticmethod
    def _return_code(process: object) -> int | None:
        poll = getattr(process, "poll", None)
        if callable(poll):
            code = poll()
            if code is not None:
                return code
        wait = getattr(process, "wait", None)
        if callable(wait):
            try:
                wait(timeout=0.2)
            except Exception:  # noqa: S110 - process may already be reaped; poll() below is authoritative
                pass
        if callable(poll):
            return poll()
        return getattr(process, "returncode", None)

    @staticmethod
    def _interrupt_or_none(hooks: AttemptHooks, *, allow_stop_only: bool) -> str | None:
        _ = allow_stop_only
        return hooks.check_interrupt()


__all__ = ["AttemptHooks", "QueueItemLike", "TrackAttemptRequest", "TrackAttemptResult", "TrackAttemptRunner"]

File: packages/usecases/src/play-track.ts
/**
 * Play one track: resolve → probe → spawn → stream chunks → classify.
 * Ported from app/usecases/play_track.py (TrackAttemptRunner).
 *
 * The runner owns ONE attempt and reports a TrackAttemptResult; the engine
 * keeps retry policy, interrupt dispatch, and state mutation behind the
 * AttemptHooks interface. All time access flows through the injected clock.
 */

import {
  ATTEMPT_COMPLETED,
  ATTEMPT_RETRY_FFMPEG,
  ATTEMPT_RETRY_SOURCE,
  classifyAttempt,
  expectedDurationSeconds,
  slowChunkRead,
} from "@airwave/domain";
import type { AttemptOutcome, ResolvedTrackLike, Transcoder } from "@airwave/domain";

export interface QueueItemLike {
  id: number;
  sourceUrl: string;
  normalizedUrl: string | null;
  durationSeconds: number | null;
  title: string | null;
}

export interface TrackAttemptRequest {
  queueItem: QueueItemLike;
  attempt: number;
  defaultSeekSeconds: number;
}

export interface TrackAttemptResult {
  outcome: AttemptOutcome;
  reason: string | null;
  chunksSent: number;
  bytesSent: number;
  elapsedSeconds: number;
  expectedSeconds: number;
  resolved: ResolvedTrackLike | null;
  seekSeconds: number;
  stderrText: string;
}

/** Engine capabilities the attempt needs. Implemented by the Node engine. */
export interface AttemptHooks {
  resolveTrack(item: QueueItemLike, forceRefresh: boolean): ResolvedTrackLike;
  onResolvedMetadata(resolved: ResolvedTrackLike): void;
  markItemResolved(itemId: number, normalizedUrl: string): void;
  rememberResolved(resolved: ResolvedTrackLike): void;
  consumeSeek(defaultSeconds: number): number;
  setPlaybackOffset(seconds: number): void;
  getPrefetchedAudio(itemId: number): string | null;
  prefetchAudio(itemId: number, sourceUrl: string): void;
  usesDirectFfmpeg(item: QueueItemLike): boolean;
  registerActiveProcess(process: unknown): void;
  triggerPrefetchUpcoming(): void;
  notifyStateChanged(): void;
  startTransitionSilence(): TransitionSilenceHandle | null;
  stopTransitionSilence(handle: TransitionSilenceHandle): void;
  onFirstChunk(): void;
  writeChunk(chunk: Buffer): void;
  /** null = keep going; "stop" ends the engine; other reasons interrupt. */
  checkInterrupt(): string | null;
  consumeInterruptReason(): string;
  /** Optional slow-read telemetry hook (engine logging policy). */
  notifySlowRead?(itemId: number, attempt: number, chunkIndex: number, readSeconds: number, requested: number, received: number): void;
}

export interface TransitionSilenceHandle {
  stop: () => void;
}

interface SilencePair {
  handle: TransitionSilenceHandle | null;
}

export class TrackAttemptRunner {
  private readonly transcoder: Transcoder;
  private readonly clock: () => number;
  private readonly chunkSize: number;
  private inResolvePhase = false;

  constructor(options: { transcoder: Transcoder; clock: () => number; chunkSize: number }) {
    this.transcoder = options.transcoder;
    this.clock = options.clock;
    this.chunkSize = options.chunkSize;
  }

  run(request: TrackAttemptRequest, hooks: AttemptHooks): TrackAttemptResult {
    const item = request.queueItem;
    const silence: SilencePair = { handle: null };
    const startedAt = this.clock();
    try {
      silence.handle = hooks.startTransitionSilence();
      const resolved = this.resolve(request, hooks);
      const probedDuration = this.probeDuration(resolved);
      hooks.markItemResolved(item.id, resolved.normalizedUrl);
      hooks.rememberResolved(resolved);
      hooks.onResolvedMetadata(resolved);
      const seekSeconds = hooks.consumeSeek(request.defaultSeekSeconds);
      hooks.setPlaybackOffset(seekSeconds);

      const process = this.spawn(request, resolved, seekSeconds, hooks);
      hooks.registerActiveProcess(process);
      hooks.triggerPrefetchUpcoming();
      hooks.notifyStateChanged();

      const { chunks, byteCount, stderrText } = this.streamChunks(item.id, request.attempt, process, hooks, () => {
        // First real chunk: stop transition silence, then engine prefetch hook.
        if (silence.handle) {
          const handle = silence.handle;
          silence.handle = null;
          hooks.stopTransitionSilence(handle);
        }
        hooks.onFirstChunk();
      });
      const elapsed = Math.max(0, this.clock() - startedAt);

      const expected = expectedDurationSeconds(probedDuration, resolved.durationSeconds, item.durationSeconds);
      const verdict = classifyAttempt({
        ffmpegReturnCode: this.returnCode(process),
        sourceReturnCode: 0,
        elapsedSeconds: elapsed,
        expectedSeconds: expected,
        stderrText,
      });
      return {
        outcome: verdict.outcome,
        reason: verdict.reason,
        chunksSent: chunks,
        bytesSent: byteCount,
        elapsedSeconds: elapsed,
        expectedSeconds: expected,
        resolved,
        seekSeconds,
        stderrText,
      };
    } catch (error) {
      if (error instanceof InterruptedError) throw error;
      const outcome = this.inResolvePhase ? ATTEMPT_RETRY_SOURCE : ATTEMPT_RETRY_FFMPEG;
      return {
        outcome,
        reason: error instanceof Error ? error.message : String(error),
        chunksSent: 0,
        bytesSent: 0,
        elapsedSeconds: Math.max(0, this.clock() - startedAt),
        expectedSeconds: 0,
        resolved: null,
        seekSeconds: 0,
        stderrText: "",
      };
    } finally {
      if (silence.handle) {
        const handle = silence.handle;
        silence.handle = null;
        hooks.stopTransitionSilence(handle);
      }
    }
  }

  // ---------------------------------------------------------------- phases

  private resolve(request: TrackAttemptRequest, hooks: AttemptHooks): ResolvedTrackLike {
    this.inResolvePhase = true;
    try {
      const resolved = hooks.resolveTrack(request.queueItem, request.attempt > 1);
      this.inResolvePhase = false;
      return resolved;
    } catch (error) {
      // Keep inResolvePhase = true so the outer catch maps this to retry_source.
      throw error;
    }
  }

  private probeDuration(resolved: ResolvedTrackLike): number | null {
    try {
      const probe = this.transcoder.probeSource(resolved.streamUrl) as
        | Record<string, number | string | null>
        | null
        | undefined;
      if (!probe) return null;
      const raw = probe["duration_seconds"] ?? probe["durationSeconds"];
      const value = typeof raw === "number" ? raw : raw !== null && raw !== undefined ? Number(raw) : NaN;
      return Number.isFinite(value) ? value : null;
    } catch {
      return null;
    }
  }

  private spawn(
    request: TrackAttemptRequest,
    resolved: ResolvedTrackLike,
    seekSeconds: number,
    hooks: AttemptHooks,
  ): unknown {
    const item = request.queueItem;
    let prefetched = hooks.getPrefetchedAudio(item.id);
    if (!prefetched && !resolved.isLive && !hooks.usesDirectFfmpeg(item)) {
      hooks.prefetchAudio(item.id, item.sourceUrl);
      const interrupt = hooks.checkInterrupt();
      if (interrupt !== null) {
        throw new InterruptedError(hooks.consumeInterruptReason());
      }
      prefetched = hooks.getPrefetchedAudio(item.id);
    }
    if (prefetched) {
      return this.transcoder.spawnForSource(prefetched, seekSeconds);
    }
    if (seekSeconds > 0 || resolved.isLive || hooks.usesDirectFfmpeg(item)) {
      return this.transcoder.spawnForSource(resolved.streamUrl, seekSeconds);
    }
    throw new Error("ffmpeg source playback is unavailable");
  }

  private streamChunks(
    itemId: number,
    attempt: number,
    process: unknown,
    hooks: AttemptHooks,
    onFirstChunk: () => void,
  ): { chunks: number; byteCount: number; stderrText: string } {
    let chunks = 0;
    let byteCount = 0;
    for (;;) {
      const interrupt = hooks.checkInterrupt();
      if (interrupt !== null) {
        throw new InterruptedError(interrupt === "stop" ? "stop" : hooks.consumeInterruptReason());
      }
      const readStarted = this.clock();
      const chunk = this.readChunk(process);
      const readSeconds = this.clock() - readStarted;
      if (chunk && slowChunkRead(readSeconds)) {
        // Slow-read warning is engine-side logging policy; surfaced via reason-free path.
        hooks.notifySlowRead?.(itemId, attempt, chunks, readSeconds, this.chunkSize, chunk.length);
      }
      if (!chunk || chunk.length === 0) {
        return { chunks, byteCount, stderrText: this.stderrText(process) };
      }
      if (chunks === 0) onFirstChunk();
      hooks.writeChunk(chunk);
      chunks += 1;
      byteCount += chunk.length;
    }
  }

  // --------------------------------------------------------------- helpers

  private readChunk(process: unknown): Buffer {
    const stdout = (process as { stdout?: { read?: (n: number) => Buffer } } | null)?.stdout;
    if (!stdout || typeof stdout.read !== "function") return Buffer.alloc(0);
    return stdout.read(this.chunkSize) ?? Buffer.alloc(0);
  }

  private stderrText(process: unknown): string {
    const readStderr = (proc: unknown): string => {
      const pipe = (proc as { stderr?: { read?: () => Buffer } } | null)?.stderr;
      if (!pipe || typeof pipe.read !== "function") return "";
      return pipe.read()?.toString("utf-8").trim() ?? "";
    };
    return readStderr(process).trim();
  }

  private returnCode(process: unknown): number | null {
    const proc = process as { poll?: () => number | null; returncode?: number | null } | null;
    if (typeof proc?.poll === "function") {
      const code = proc.poll();
      if (code !== null && code !== undefined) return code;
    }
    return proc?.returncode ?? null;
  }
}

/** Mirrors Python's InterruptedError control-flow contract. */
export class InterruptedError extends Error {
  readonly reason: string;

  constructor(reason: string) {
    super(reason);
    this.name = "InterruptedError";
    this.reason = reason;
  }
}

export { ATTEMPT_COMPLETED, ATTEMPT_RETRY_FFMPEG, ATTEMPT_RETRY_SOURCE };
